src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -0,0 +1,256 @@
+
+        #endregion
+
+        /// <summary>
+        /// constructor for BatchSaveResult
+        /// </summary>
+        /// <param name="context">context</param>
+        /// <param name="method">method</param>
+        /// <param name="options">options</param>
+        /// <param name="callback">user callback</param>
+        /// <param name="state">user state object</param>
+        internal BulkUpdateSaveResult(DataServiceContext context, string method, SaveChangesOptions options, AsyncCallback callback, object state)
+            : base(context, method,null, options, callback, state)
+        {
+            Debug.Assert(Util.IsBulkUpdate(options), "the options must have bulk update flag set");
+        }
+
+        protected override Stream ResponseStream => throw new NotImplementedException();
+
+        /// <summary>
+        /// Synchronous bulk update request
+        /// </summary>
+        internal void BulkUpdateRequest<T>(params T[] objects)
+        {
+            List<Descriptor> descriptorGraph = ProcessDescriptors(this.ChangedEntries, objects);
+            ODataRequestMessageWrapper bulkUpdateRequestMessage = this.GenerateBulkUpdateRequest(descriptorGraph);
+
+            if (bulkUpdateRequestMessage != null)
+            {
+                bulkUpdateRequestMessage.SetRequestStream(bulkUpdateRequestMessage.CachedRequestStream);
+
+                try
+                {
+                    this.batchResponseMessage = this.RequestInfo.GetSynchronousResponse(bulkUpdateRequestMessage, false);
+                }
+                catch (DataServiceTransportException ex)
+                {
+                    InvalidOperationException exception = WebUtil.GetHttpWebResponse(ex, ref this.batchResponseMessage);
+
+                    throw exception;
+                }
+                finally
+                {
+                    if (this.batchResponseMessage != null)
+                    {
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// This method processes the descriptors. It loops through
+        /// all the descriptors and creates a graph-like structure of 
+        /// the descriptors and their relationships It uses the objects 
+        /// provided by the customer as the root descriptors
+        /// </summary>
+        /// <typeparam name="T">The type of the top-level resources we want to do a deep update on. </typeparam>
+        /// <param name="descriptors">The list of descriptors in the entity tracker.</param>
+        /// <param name="objects">Objects of the top-level type we want to perform a deep update on. </param>
+        /// <returns> A list of the processed descriptors. This list will contain the descriptors in a graph-like manner.</returns>
+        /// <exception cref="InvalidOperationException"></exception>
+        public List<Descriptor> ProcessDescriptors<T>(IEnumerable<Descriptor> descriptors, params T[] objects)

```

Comment:
Does this method traverse all the descriptors in the entity tracker? Does it do that each time you make a bulk request? Would that not be expensive if the entity tracker is large? 

Concurrent requests might me be calling this method at the same time, and since it mutates the descriptors (i.e., adding related descriptors and changing the entity set name), it's not thread safe.

Is it possible to build this graph once and re-use it, and only update it if links between entities change?

Alternatively, maybe the internal data structures of the entity tracker could be changed to better represent the graph, such that when you add an entity or a link, the graph edges are created or updated at that point?